### PR TITLE
 Finish implementing JsonSchema[OpenApi] instance.

### PR DESCRIPTION
**Pull request by julienrf**
- Remove previous circe encoder from OpenApi model
- Remove circe dependency from `openapi` and `example-documented`
  modules
- Align circe encoding of optional fields with play-json: a `None`
  value is encoded with a missing field rather than a `null` field
- All Schemas can now have an attached “description”

Fixes julienrf#126

### CODE REVIEW

**Comment by julienrf** [comment]
The diff is huge because I’ve merged `master` into the feature branch first. Only the last commit is relevant.

File: json-schema/json-schema-circe/src/main/scala/endpoints/circe/JsonSchemas.scala [hunk 1]
```diff
@@ -113,7 +113,7 @@ trait JsonSchemas
   // FIXME Check that this is the correct way to model optional fields with circe
   def optField[A](name: String, documentation: Option[String] = None)(implicit tpe: JsonSchema[A]): Record[Option[A]] =
     Record(
-      io.circe.ObjectEncoder.instance[Option[A]](a => JsonObject.singleton(name, io.circe.Encoder.encodeOption(tpe.encoder).apply(a))),
+      io.circe.ObjectEncoder.instance[Option[A]](maybeA => JsonObject.fromIterable(maybeA.map(a => name -> tpe.encoder.apply(a)))),
```
Comment (julienrf) [inline json-schema/json-schema-circe/src/main/scala/endpoints/circe/JsonSchemas.scala:116]: This is an important change: I have aligned the behavior of the circe encoders with the behavior of the play-json encoders for optional fields: an optional field whose value is `None` is **not emitted** in the resulting JSON object (it was previously emitted with a `null` value).

File: json-schema/json-schema-circe/src/main/scala/endpoints/circe/JsonSchemas.scala [hunk 2]
```diff
@@ -144,7 +144,10 @@ trait JsonSchemas
   def zipRecords[A, B](recordA: Record[A], recordB: Record[B]): Record[(A, B)] = {
     val encoder =
       io.circe.ObjectEncoder.instance[(A, B)] { case (a, b) =>
-        recordA.encoder.apply(a).deepMerge(recordB.encoder.apply(b)).asObject.get
+        // For some reason, `deepMerge` puts the fields of its left-hand-side *after*
+        // the fields of its right-hand-side. Hence the inversion between `recordA`
+        // and `recordB`.
+        recordB.encoder.apply(b).deepMerge(recordA.encoder.apply(a)).asObject.get
```
Comment (julienrf) [inline json-schema/json-schema-circe/src/main/scala/endpoints/circe/JsonSchemas.scala:150]: This change produces a prettier output when we rely on fields ordering.

File: openapi/openapi/src/test/scala/endpoints/openapi/ReferencedSchemaTest.scala [hunk 3]
```diff
@@ -36,179 +42,201 @@ class ReferencedSchemaTest extends WordSpec with Matchers {
...
+        |    },
+        |    "securitySchemes" : {
+        |      "HttpBasic" : {
+        |        "type" : "http",
+        |        "description" : "Http Basic Authentication",
+        |        "scheme" : "basic"
+        |      }
+        |    }
+        |  }
+        |}""".stripMargin
```
Comment (julienrf) [inline openapi/openapi/src/test/scala/endpoints/openapi/ReferencedSchemaTest.scala:222]: The produced schema is the same as before, but the order of some field has changed, and I have added a case where we have a property with a `description`.

**Comment by julienrf** [comment]
Note that the definition of the `JsonSchema[OpenApi]` is partial: it only works for JSON encoders, not JSON decoders. This means that we can not decode Open API documents. It will fail at runtime, although the code will compile, if you try…

File: json-schema/json-schema-circe/src/main/scala/endpoints/circe/JsonSchemas.scala (hunk 1 above)
Comment (stefanholzmueller) [reply json-schema/json-schema-circe/src/main/scala/endpoints/circe/JsonSchemas.scala:116]: Thanks! This will also help with #123

**Comment by stefanholzmueller** [comment]
> Note that the definition of the `JsonSchema[OpenApi]` is partial: it only works for JSON encoders, not JSON decoders. This means that we can not decode Open API documents. It will fail at runtime, although the code will compile, if you try…

I cannot find the place in the code where it will crash. Anyways, it is not a big deal for me.

**Comment by stefanholzmueller** [comment]
@julienrf Thanks for completing this PR! Unfortunately, in the past weeks I have not had the time for OS work. Maybe that will change this year, but I don't want to promise anything I cannot keep.

I saw a lot of `TODO`s in the diff, not sure if you are comfortable merging them into julienrf/endpoints. I'll just merge your PR into my PR and then let's see.

**Comment by julienrf** [comment]
Thanks!

**Comment by julienrf** [comment]
I’ve created https://github.com/julienrf/endpoints/pull/214 on the origin repo.

**Pull request merged**